src/Util/CommandRunner.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace SilverStripe\Upgrader\Util;
+
+use SilverStripe\Upgrader\ChangeDisplayer;
+use SilverStripe\Upgrader\CodeCollection\CodeChangeSet;
+use SilverStripe\Upgrader\Console\AbstractCommand;
+use SilverStripe\Upgrader\Console\AutomatedCommand;
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Exception\LogicException;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Aggregate many commands together and run them sequentially.
+ */
+class CommandRunner
+{
+
+    /**
+     * Run a list of command sequentially.
+     * @param Application $app
+     * @param array $commands
+     * @param array $args
+     * @param OutputInterface $output
+     */
+    public function run(
+        Application $app,
+        array $commands,
+        array $args,
+        OutputInterface $output
+    ): void {
+        $args = $this->addProjectFolders($args);
+
+        $step = 0;
+
+        $warnings = new CodeChangeSet();
+
+        // Run each command one by one.
+        foreach ($commands as $commandName) {
+            $step++;
+
+            /**
+             * @var AutomatedCommand|AbstractCommand
+             */
+            $cmd = $app->find($commandName);
+            if ($cmd instanceof AutomatedCommand) {
+                $titleBlock = sprintf('Step %s - Running %s', $step, $cmd->getName());
+
+                $output->write($this->wrapTitle($titleBlock));
+                $cmd->automatedRun($args, $output);
+                $args = $cmd->updatedArguments();
+                $diff = $cmd->getDiff();
+                if ($diff) {
+                    $warnings->mergeWarnings($diff);
+                }
+            } else {
+                throw new LogicException(sprintf(
+                    '%s does not implement the AutomatedCommand interface.',
+                    $commandName
+                ));
+            }
+        }
+
+        // Show a summary of the upgrade
+        $output->write($this->wrapTitle('Summary of upgrades'));
+
+        $output->writeln('All commands got executed successfully. Please address any outstanding warnings.');
+
+        $displayer = new ChangeDisplayer();
+        $displayer->displayWarningsOnly($output, $warnings);
+    }
+
+    /**
+     * Add a `project-path` key and `code-path` key to the arguments. Commands who requires a `path` argument can pick
+     * one of those two values.
+     *
+     * @internal This is necessary because some commands expect to be run on the `mysite/code` folder, while others
+     * expect to be run on `mysite`.
+     * @param array $args
+     * @return array
+     */
+    public function addProjectFolders(array $args): array
+    {
+        if (isset($args['--root-dir'])) {
```

Comment:
Since all of these source arguments are actually set by the AllInOneCommand it feels disjoint to rely on these in a separate class. Perhaps we should shift this logic into the command class instead? That way your argument definition and normalisation are consolidated.

Reply:
Had a bit of a chat with  @tractorcow about this one. I think I'll leave that bit of code here. It could be helpful to move it up to the actual command, although it could complicate things a little bit down the road if we decide to do a dedicate command for upgrading packages.

It works for now, but it might be worth revisit it later down the road.